javascript/extractor/src/com/semmle/js/extractor/AutoBuild.java

```diff
@@ -444,33 +449,119 @@ protected boolean hasSeenCode() {
 
   /** Perform extraction. */
   public int run() throws IOException {
-    startThreadPool();
-    try {
-      CompletableFuture<?> sourceFuture = extractSource();
-      sourceFuture.join(); // wait for source extraction to complete
-      if (hasSeenCode()) { // don't bother with the externs if no code was seen
-        extractExterns();
+    initDiagnosticsWriter();
+
+    try (DiagnosticWriter writerToClose = diagnostics) {
```

Comment:
Is the DiagnosticsWriter thread-safe?

Reply:
Nope, nicely spotted. 
I'll use `ThreadLocal` to make one writer pr. thread.